keps/sig-api-machinery/3158-optional-maximum-limit-on-list-requests-to-etcd/README.md

```diff
@@ -0,0 +1,901 @@
+Trace[638277422]: [2.272115626s] [2.272115626s] END
+```
+10x clients issued such queries lead to etcd OOM-killed and failed other APIServer requests hitting to that node.
+
+With the approach and `max-list-etcd-limit = 500`, user should expect 
+- APIServer is healthy and serving requests because etcd cluster is healthy 
+- List etcd duration increases from 2 seconds around 3 seconds and this duration is available as a prometheus metric exposed in "/metrics" endpoint. 
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional) 
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+- Setting the maximum limit too low will increase the total round-trip time between APIServer and etcd to breach the default 1 minute request time. This will cause APIServer cancelles the requests prematurely. Example https://github.com/kubernetes/kubernetes/pull/94303. 
+
+- Use the maximum limit as suggestion and adjust it dynamically to fit in the 1 minute request timeout. For example, assuming each resource object size is identical given resource type, if the very first list with suggested maximum limit takes 2s but requires another 60 such subsequent list, double the limit to reduce the round trip time. 
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+Modify [apiserver/blob/master/pkg/storage/etcd3/store.go](https://github.com/kubernetes/apiserver/blob/master/pkg/storage/etcd3/store.go#L565-L576) list to apply a maximum limit to every etcd range request and issue multiple range requests if necessary to satisfy the original list. Filtering and decoding on each etcd response before issuing the next range request. Using the ResourceVersion that comes back from the first range request from etcd (or the one user provided) on later requests to ensure a consistent snapshot of resource collections. The implementation should unify with [current pagination implementation](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go#L686)
+
+The maximum limit will be a per-resource-type limit and exposed as an optional APIServer flag. The current behavior will not be changed by default.
```

Comment:
change to -> `The new default behavior will continue to be the same as before when the flag is unspecified, i.e no max page-size will be imposed on the list calls to etcd`